### PR TITLE
Output errors in compliance with RFC7807

### PR DESCRIPTION
**Pull request by eranmes**
Change the error output format to be the Problem Details Object from
RFC7807.

Seems pretty straightforward given that the 6962-bis errors do not
have unique requirements and so can comply with said RFC.

### CODE REVIEW

File: draft-ietf-trans-rfc6962-bis.md [hunk 1]
```diff
@@ -1179,8 +1185,9 @@ following:
 
 ~~~~~~~~~~~
     {
-        "error_message": "'start' cannot be greater than 'end'",
-        "error_code": "not compliant",
+        "type": "https://example.net/get-entries-error"
+        "title": "not compliant",
+        "detail": "'start' cannot be greater than 'end'",
```
Comment (robstradling) [inline draft-ietf-trans-rfc6962-bis.md]: Using URNs (a la ACME), "not compliant" would become a camelCase suffix to the "type" parameter.  i.e.,

```
{
    "type": "urn:ietf:params:ct:error:notCompliant",
    "detail": "'start' cannot be greater than 'end'"
}
```

A few other thoughts...
"notCompliant" seems unnecessarily generic.  For problems with request parameters, ACME uses "malformed" and RFC7807 suggests "validation-error".
RFC7807 has the concept of subproblems, permitting details of multiple similar problems to be conveyed.  ACME does this.  Using https://tools.ietf.org/html/draft-ietf-acme-acme-14#section-6.6.1 and https://tools.ietf.org/html/rfc7807#section-3 for inspiration, this example might look something like this...

```
{
    "type": "urn:ietf:params:ct:error:malformed",
    "detail": "The request message was malformed",
    "subproblems": [
        {
            "name": "start",
            "detail": "must not be greater than 'end'"
        },
        {
            "name": "end",
            "detail": "must not be less than 'start'"
        }
    ]
}
```

File: draft-ietf-trans-rfc6962-bis.md [hunk 2]
```diff
@@ -1151,24 +1152,29 @@ requirement above).
...
 
-error_message:
+type:
+: A URI reference identifying the problem. For example, a reference to this
+  document.
+
+title:
+: An error code readable by the client. Other than the generic codes detailed
+  here, each error code is specific to the type of request. Specific
+  errors are specified in the respective sections below. Error codes are fixed
```
[the remark this answers is not in the document]
Comment (eranmes) [reply draft-ietf-trans-rfc6962-bis.md]: Done.

File: draft-ietf-trans-rfc6962-bis.md [hunk 3]
```diff
@@ -1179,8 +1185,9 @@ following:
 
 ~~~~~~~~~~~
     {
-        "error_message": "'start' cannot be greater than 'end'",
-        "error_code": "not compliant",
+        "type": "https://example.net/get-entries-error"
```
[the remark this answers is not in the document]
Comment (eranmes) [reply draft-ietf-trans-rfc6962-bis.md]: Done.

File: draft-ietf-trans-rfc6962-bis.md (hunk 1 above)
Comment (eranmes) [reply draft-ietf-trans-rfc6962-bis.md]: (1) Switched to 'malformed' as suggested.
(2) Not sure the 'subproblems' structure makes a lot of sense: All but one of the errors would fall into the 'malformed' category (the exception being 'shutdown' which is a bad log state).

**Comment by eranmes** [comment]
@robstradling PTAL.

File: draft-ietf-trans-rfc6962-bis.md [hunk 4]
```diff
@@ -1147,26 +1148,29 @@ requirement above).
...
 
-error_code:
-: An error code readable by the client. Other than the generic codes detailed
-  here, each error code is specific to the type of request. Specific
-  errors are specified in the respective sections below. Error codes are fixed
-  text strings.
 
 |---------------+---------------------------------------------|
-| Error Code    | Meaning                                     |
+| Type          | Detail                                      |
```
Comment (robstradling) [inline draft-ietf-trans-rfc6962-bis.md]: I think "type" and "detail" (i.e., no capitals) would be less confusing column headers for these tables.

File: draft-ietf-trans-rfc6962-bis.md [hunk 5]
```diff
@@ -1147,26 +1148,29 @@ requirement above).
...
-
-error_message:
+outlined in the subsections below, the body SHOULD be a JSON Problem Details
+Object (see [RFC7807] Section 3), containing:
+
+type:
+: A URN reference identifying the problem. To facilitate automated diagnostics
+  of errors, this document defines a set of standard tokens for use in the
+  `type` field, within the URN namespace of: "urn:ietf:params:trans:error:".
+  Other than the generic codes detailed here, each error type is specific to
```
Comment (robstradling) [inline draft-ietf-trans-rfc6962-bis.md]: We're only specifying one generic error: "malformed".

I suggest: s/the generic codes detailed here/the generic "malformed" error type defined in this section/

File: draft-ietf-trans-rfc6962-bis.md [hunk 6]
```diff
@@ -1147,26 +1148,29 @@ requirement above).
...
-  here, each error code is specific to the type of request. Specific
-  errors are specified in the respective sections below. Error codes are fixed
-  text strings.
 
 |---------------+---------------------------------------------|
-| Error Code    | Meaning                                     |
+| Type          | Detail                                      |
 |---------------+---------------------------------------------|
-| not compliant | The request is not compliant with this RFC. |
+| malformed     | The request is not compliant with this RFC. |
```
Comment (robstradling) [inline draft-ietf-trans-rfc6962-bis.md]: This table seems to be saying that "detail" will be always be the fixed string "The request is not compliant with this RFC", which obviously conflicts with the previous sentence (i.e., "...SHOULD provide sufficient detail for the error to be rectified").

How about changing "The request is not compliant with this RFC" to "\<sufficient detail for the error to be rectified>"?

File: draft-ietf-trans-rfc6962-bis.md [hunk 7]
```diff
@@ -1147,26 +1148,29 @@ requirement above).
...
 response code of 4xx/5xx (see [RFC7231]), and, in place of the responses
-outlined in the subsections below, the body SHOULD be a JSON structure
-containing at least the following field:
-
-error_message:
+outlined in the subsections below, the body SHOULD be a JSON Problem Details
+Object (see [RFC7807] Section 3), containing:
+
+type:
+: A URN reference identifying the problem. To facilitate automated diagnostics
```
[the remark this answers is not in the document]
Comment (eranmes) [reply draft-ietf-trans-rfc6962-bis.md]: Sounds reasonable, done.

File: draft-ietf-trans-rfc6962-bis.md (hunk 5 above)
Comment (eranmes) [reply draft-ietf-trans-rfc6962-bis.md]: Done.

File: draft-ietf-trans-rfc6962-bis.md (hunk 4 above)
Comment (eranmes) [reply draft-ietf-trans-rfc6962-bis.md]: Done, here and throughout.

File: draft-ietf-trans-rfc6962-bis.md (hunk 6 above)
Comment (eranmes) [reply draft-ietf-trans-rfc6962-bis.md]: Per OOB discussion:
* Left 'malformed' as a general error.
* Specified type and detail for each of the cases where 'malformed' might have been used.

**Comment by eranmes** [comment]
@robstradling PTAL.

**Comment by robstradling** [comment]
Thanks @eranmes.  I'm going to merge this and then propose a follow-on PR to clean up a couple of things.

**Comment by robstradling** [comment]
My follow-on PR is #301.

**Pull request merged**